### ISSUE TITLE
clang flase-positive implicit declaration

### ISSUE THREAD

**Issue by JAicewizard** [llvm/llvm-project#55489]
I am fairly certain that I have found a false positive in clang, using the latest from the arch Linux repos.

```console
$ clang --version
clang version 13.0.1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /sbin
```

Example program:
```c
#include <stdio.h>

int main() {
	char password[256];
	gets(password);
}
```

clang output when compiling file above:
```console
main.c:5:2: warning: implicit declaration of function 'gets' is invalid in C99 [-Wimplicit-function-declaration]
        gets(password);
        ^
1 warning generated.
/sbin/ld: /tmp/main-88b226.o: in function `main':
main.c:(.text+0x22): warning: the `gets' function is dangerous and should not be used.
```

This is not an implicit declaration of `gets`, as I imported `stdio.h` and `gets` is defined in this header per the standard.

**Comment by shafik** [comment]
`gets()` was removed in C11 if you use `-std=c99` you will see the following diagnostic:

```console
 warning: 'gets' is deprecated [-Wdeprecated-declarations]
        gets(password);
        ^
```

[see godbolt example](https://godbolt.org/z/h6jvbhKz1)

**Comment by JAicewizard** [comment]
In that case clang misinforms me about the standard which it is using, as the warning lists C99 not C11.

**Comment by shafik** [comment]
Implicit function declarations was removed in C99, which is what the diagnostic is saying `is invalid in C99` and forward. If use [clang trunk](https://godbolt.org/z/xP9bc5jo8) it is more explicit about this:

```console
call to undeclared function 'gets'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
```

**Comment by JAicewizard** [comment]
In that case the message is very unclear about its actual meaning. But since this has changed already I think this issue can be closed.
Thank you for your quick responses!
